## Text Analytics Group Assignment 2
### ArjunVarma_AnyingLi_YuanzhuoPeng_JiaqiuWang_NimishAmlathe

In [1]:
from pandas import Series, DataFrame
import pandas as pd
from patsy import dmatrices
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [2]:
yelp = pd.read_csv('Yelp Data Restaurant Reviews Ratings.csv')
yelp['rating'] = yelp['stars'].map(lambda x: 'High' if x >3  else 'Low')
yelp[:5]

stars  votes_cool  votes_funny  votes_useful  Cheap  Moderate  Expensive  \
0      1           0            0             0      1         0          0   
1      2           2            2             2      1         0          0   
2      4           0            0             1      1         0          0   
3      3           0            1             2      1         0          0   
4      3           7            9             9      1         0          0   

   VeryExpensive  American  Chinese   ...    Indian  Italian  Greek  \
0              0         0        0   ...         0        0      0   
1              0         0        0   ...         0        0      0   
2              0         0        0   ...         0        0      0   
3              0         0        0   ...         0        0      0   
4              0         0        0   ...         0        0      0   

   Mediterranean  Mexican  Thai  Vietnamese  Others  \
0              0        0     0           0       1   
1              0        0     0           0       1   
2              0        0     0           0       1   
3              0        0     0           0       1   
4              0        0     0           0       1   

                                              Review  rating  
0  This location is out of business. I drove by i...     Low  
1  = = = = = = CLOSED = = = = = =This JB s locati...     Low  
2  This is just a basic (albeit mini) chain greas...    High  
3  Whenever I offer to take my mom out to lunch s...     Low  
4  If I say it wasn t as bad as I was expecting i...     Low  

[5 rows x 22 columns]

In [3]:
yelp.columns.values

array(['stars', 'votes_cool', 'votes_funny', 'votes_useful', 'Cheap',
       'Moderate', 'Expensive', 'VeryExpensive', 'American', 'Chinese',
       'French', 'Japanese', 'Indian', 'Italian', 'Greek', 'Mediterranean',
       'Mexican', 'Thai', 'Vietnamese', 'Others', 'Review', 'rating'], dtype=object)

#### Task A. Ignore the text (reviews) and run a classification model with the numeric data (you can use standard methods like logistic regression, k-nearest neighbors or anything else). What is the best accuracy of your model with numeric data?

In [4]:
# Take only a random subset of 2000 rows
yelp_subset = yelp.sample(frac=.1,random_state=5)
yelp_subset = yelp_subset.reset_index()
yelp_subset = yelp_subset.drop('index',axis=1)
yelp_subset[:3]

stars  votes_cool  votes_funny  votes_useful  Cheap  Moderate  Expensive  \
0      5           0            0             0      0         1          0   
1      2           0            1             3      0         0          1   
2      5           0            1             1      0         1          0   

   VeryExpensive  American  Chinese   ...    Indian  Italian  Greek  \
0              0         1        0   ...         0        0      0   
1              0         1        0   ...         0        0      0   
2              0         1        0   ...         0        0      0   

   Mediterranean  Mexican  Thai  Vietnamese  Others  \
0              0        0     0           0       0   
1              0        0     0           0       0   
2              0        0     0           0       0   

                                              Review  rating  
0  I ll be honest...I m not a food connoisseur  b...    High  
1  Made reservations for 5:30 PM on a Saturday ni...     Low  
2  This place has an awesome Happy Hour. The envi...    High  

[3 rows x 22 columns]

In [5]:
#split train and test data
from sklearn.cross_validation import StratifiedShuffleSplit 

y = yelp_subset['rating'].map(lambda y: 1 if y == 'High'  else 0)

X = yelp_subset.drop(['stars','Review','rating'], axis=1)

Review = yelp_subset['Review']

sss = StratifiedShuffleSplit(y, test_size=0.3, random_state=1)
for train_index, test_index in sss:
    break

train_x, train_y,train_review = X.values[train_index],y.values[train_index],Review[train_index]
test_x, test_y, test_review = X.values[test_index], y.values[test_index],Review[test_index]

In [6]:
#knn model with numerical inputs only
from sklearn import neighbors
modelknn = neighbors.KNeighborsClassifier(n_neighbors=15, weights='uniform', p=2)
modelknn.fit(train_x,train_y)

from sklearn import metrics
prediction_on_training = modelknn.predict(train_x)
print "Train set accuracy score:", metrics.accuracy_score(train_y, prediction_on_training)

predicted_classes = modelknn.predict(test_x)
print "Test set accuracy score:", metrics.accuracy_score(test_y, predicted_classes)

Train set accuracy score: 0.695714285714
Test set accuracy score: 0.703333333333


In [7]:
#logistic regression with numerical inputs only
from sklearn.linear_model import LogisticRegression
modellog = LogisticRegression()
modellog.fit(train_x, train_y)

from sklearn import metrics
prediction_on_training = modellog.predict(train_x)
print "Train set accuracy score:", metrics.accuracy_score(train_y, prediction_on_training)

predicted_classes = modellog.predict(test_x)
print "Test set accuracy score:", metrics.accuracy_score(test_y, predicted_classes)

Train set accuracy score: 0.692857142857
Test set accuracy score: 0.701666666667


In [8]:
#naive bayes with numerical inputs only
from sklearn import naive_bayes
modelnby = naive_bayes.MultinomialNB()
modelnby.fit(train_x, train_y)

from sklearn import metrics
prediction_on_training = modelnby.predict(train_x)
print "Train set accuracy score:", metrics.accuracy_score(train_y, prediction_on_training)

predicted_classes = modelnby.predict(test_x)
print "Test set accuracy score:", metrics.accuracy_score(test_y, predicted_classes)

Train set accuracy score: 0.687142857143
Test set accuracy score: 0.686666666667


So the best accuracy with using only numeric data for prediction is about 70.3% on test data with k nearest neighbor model.

#### Task B. Perform a supervised classification on a subset of the corpus using the reviews only. You can write your code in Python or R. What accuracy do you get from this text mining exercise?

In [9]:
import operator
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer


import re
import string
import nltk
from nltk import word_tokenize          
from nltk.stem import PorterStemmer, WordNetLemmatizer
from nltk.corpus import stopwords


stemmer = PorterStemmer()
wnl = WordNetLemmatizer()
def stem_tokens(tokens, stemmer):
    stemmed = []
    for item in tokens:
        stemmed.append(stemmer.stem(item))
    return stemmed

def lemmatize(tokens,wnl):
    lem=[]
    for item in tokens:
        lem.append(wnl.lemmatize(item))
    return lem

def my_tokenizer(text):
    text = "".join([ch for ch in text if ch not in string.punctuation])
    text = re.sub('[0-9]+','',text)
    tokens = nltk.word_tokenize(text)
    tokens = [word for word in tokens if word not in stopwords.words('english')]
    #result = stem_tokens(tokens, stemmer)
    #result = lemmatize(tokens, wnl)
    result = tokens
    return result
    

vectorizer = TfidfVectorizer(min_df=3,max_df=0.9,tokenizer=my_tokenizer,
 ngram_range=(1, 2), 
 stop_words='english',
 strip_accents='unicode',max_features=200)

text_1 = vectorizer.fit_transform(train_review)

In [10]:
from nltk import NaiveBayesClassifier
from sklearn import metrics
from sklearn.naive_bayes import MultinomialNB

text_2 = vectorizer.transform(test_review)

nb =  MultinomialNB().fit(text_1, train_y)

y_nb_predicted = nb.predict(text_2)


print "MODEL: Multinomial Naive Bayes\n"

print 'The accuracy for this classifier is ' + str(metrics.accuracy_score(test_y, y_nb_predicted))

MODEL: Multinomial Naive Bayes

The accuracy for this classifier is 0.72


So the accuracy with using only review data for prediction is about 72% on test data with Multinomial Naive Bayes Model.

#### Task C. Combine the numeric data and the text classification model (in task B) to create a “hybrid” model. It is your task to figure out how to do this. Now run this hybrid classification model and compare the results with those in A and B. Does the numeric data add to the predictive power relative to text?

In [11]:
from sklearn.naive_bayes import MultinomialNB
from operator import itemgetter

N = 50
vocabulary = np.array([t for t, i in sorted(vectorizer.vocabulary_.iteritems(), key=itemgetter(1))])

topN = np.argsort(nb.coef_[0])[-N:]

print "\nThe top %d most informative features: \n%s" % (N, ','.join(vocabulary[topN]))


The top 50 most informative features: 
perfect,times,hour,meal,people,rice,right,wait,tacos,mexican,got,sauce,salad,definitely,order,burger,dinner,eat,atmosphere,excellent,cheese,staff,bar,favorite,ordered,phoenix,happy,pizza,awesome,friendly,lunch,try,fresh,little,nice,amazing,restaurant,menu,delicious,chicken,time,really,best,love,like,service,place,good,food,great


In [12]:
most_informative = vocabulary[topN]

for w in most_informative:
    X[w]= 0

def create_features(tokens,i):
    document_words = set(tokens)
    features = {}
    for word in most_informative:
        if word in document_words:
            X.loc[i,word] = 1
        
X.head()

votes_cool  votes_funny  votes_useful  Cheap  Moderate  Expensive  \
0           0            0             0      0         1          0   
1           0            1             3      0         0          1   
2           0            1             1      0         1          0   
3           0            0             0      0         0          0   
4           0            0             0      0         1          0   

   VeryExpensive  American  Chinese  French  ...    time  really  best  love  \
0              0         1        0       0  ...       0       0     0     0   
1              0         1        0       0  ...       0       0     0     0   
2              0         1        0       0  ...       0       0     0     0   
3              1         1        0       0  ...       0       0     0     0   
4              0         1        0       0  ...       0       0     0     0   

   like  service  place  good  food  great  
0     0        0      0     0     0      0  
1     0        0      0     0     0      0  
2     0        0      0     0     0      0  
3     0        0      0     0     0      0  
4     0        0      0     0     0      0  

[5 rows x 69 columns]

In [13]:
for i in range(1000):
    text = "".join([ch for ch in Review[i] if ch not in string.punctuation])
    text = re.sub('[0-9]+','',text)
    tokens = nltk.word_tokenize(text.lower().decode('utf-8'))
    tokens = [word for word in tokens if word not in stopwords.words('english')]
    feature = create_features(tokens,i)

X.head()

votes_cool  votes_funny  votes_useful  Cheap  Moderate  Expensive  \
0           0            0             0      0         1          0   
1           0            1             3      0         0          1   
2           0            1             1      0         1          0   
3           0            0             0      0         0          0   
4           0            0             0      0         1          0   

   VeryExpensive  American  Chinese  French  ...    time  really  best  love  \
0              0         1        0       0  ...       0       0     0     1   
1              0         1        0       0  ...       1       0     1     0   
2              0         1        0       0  ...       0       0     0     0   
3              1         1        0       0  ...       0       1     0     0   
4              0         1        0       0  ...       0       0     0     0   

   like  service  place  good  food  great  
0     0        0      0     1     1      0  
1     1        1      1     1     1      1  
2     0        0      1     1     1      0  
3     0        1      0     0     0      0  
4     0        0      0     0     0      0  

[5 rows x 69 columns]

In [14]:
nb_text =  MultinomialNB().fit(X.values[train_index], train_y)

text_predicted = nb_text.predict(X.values[test_index])


print "Hybrid MODEL: Multinomial Naive Bayes\n"

print 'The accuracy for this classifier is ' + str(metrics.accuracy_score(test_y, text_predicted))

Hybrid MODEL: Multinomial Naive Bayes

The accuracy for this classifier is 0.723333333333


In [15]:
sample_weight = np.random.rand(y.shape[0])

sw_train, sw_test = sample_weight[train_index], sample_weight[test_index]

from sklearn.calibration import CalibratedClassifierCV

from sklearn.metrics import brier_score_loss


clf = MultinomialNB()
model_original = clf.fit(X.values[train_index], train_y)  # GaussianNB itself does not support sample-weights
prob_pos_clf = clf.predict_proba(X.values[test_index])[:, 1]
clf_predicted =model_original.predict(X.values[test_index])
print 'The accuracy for this classifier is ' + str(metrics.accuracy_score(test_y, clf_predicted))

# Naive-Bayes with isotonic calibration
clf_isotonic = CalibratedClassifierCV(clf, cv=10, method='isotonic')
model_isotonic = clf_isotonic.fit(X.values[train_index], train_y, sw_train)
prob_pos_isotonic = clf_isotonic.predict_proba(X.values[test_index])[:, 1]
isotonic_predicted = model_isotonic.predict(X.values[test_index])
print 'The accuracy for this classifier is ' + str(metrics.accuracy_score(test_y, isotonic_predicted))

# Naive-Bayes with sigmoid calibration
clf_sigmoid = CalibratedClassifierCV(clf, cv=10, method='sigmoid')
model_sigmoid  = clf_sigmoid.fit(X.values[train_index], train_y, sw_train)
prob_pos_sigmoid = clf_sigmoid.predict_proba(X.values[test_index])[:, 1]
sigmoid_predicted = model_sigmoid.predict(X.values[test_index])
print 'The accuracy for this classifier is ' + str(metrics.accuracy_score(test_y, sigmoid_predicted))

print("Brier scores: (the smaller the better)")

clf_score = brier_score_loss(test_y, prob_pos_clf, sw_test)
print("No calibration: %1.3f" % clf_score)

clf_isotonic_score = brier_score_loss(test_y, prob_pos_isotonic, sw_test)
print("With isotonic calibration: %1.3f" % clf_isotonic_score)

clf_sigmoid_score = brier_score_loss(test_y, prob_pos_sigmoid, sw_test)
print("With sigmoid calibration: %1.3f" % clf_sigmoid_score)

The accuracy for this classifier is 0.723333333333
The accuracy for this classifier is 0.716666666667
The accuracy for this classifier is 0.706666666667
Brier scores: (the smaller the better)
No calibration: 0.196
With isotonic calibration: 0.196
With sigmoid calibration: 0.200


#### Task D. Use unsupervised sentiment analysis on the reviews (with SentiStrength or any other tool) and use the sentiment scores to predict high/low rating. Compare and contrast the results of tasks B and D. What can you conclude from your analysis?

In [46]:
# Export yelp_subset
yelp_subset.to_csv('yelp_subset.csv',index=False)

# Run sentistrength on column 21 of yelp_subset to get positive and negative sentiments of each review
# Output from sentistrength: yelp_subset_senti+results.txt

# Load back into python
yelp_senti = pd.read_table('yelp_subset_senti+results.txt')
yelp_senti['Overall_senti'] = yelp_senti['Postive_senti'] + yelp_senti['Negative_senti']

# Get predicted ratings
yelp_senti["Pred_rating"] = np.where((yelp_senti['Overall_senti'] > 0),"High","Low")
yelp_senti[:3]

stars  votes_cool  votes_funny  votes_useful  Cheap  Moderate  Expensive  \
0      2           0            1             3      0         0          1   
1      5           0            1             1      0         1          0   
2      3           0            0             0      0         0          0   

   VeryExpensive  American  Chinese     ...       Mexican  Thai  Vietnamese  \
0              0         1        0     ...             0     0           0   
1              0         1        0     ...             0     0           0   
2              1         1        0     ...             0     0           0   

   Others                                             Review  Rating_final  \
0       0  Made reservations for 5:30 PM on a Saturday ni...           Low   
1       0  This place has an awesome Happy Hour. The envi...          High   
2       0  Noted on the OT reservation it was hubby s 50t...           Low   

   Postive_senti  Negative_senti  Overall_senti  Pred_rating  
0              3              -4             -1          Low  
1              3              -1              2         High  
2              4              -3              1         High  

[3 rows x 26 columns]

In [48]:
# Checking accuracy
from sklearn.metrics import confusion_matrix
y_true = yelp_senti['Rating_final'].tolist()
y_pred = yelp_senti['Pred_rating'].tolist()
cm_sk = confusion_matrix(y_true, y_pred)

cm = nltk.ConfusionMatrix(y_true, y_pred)
print (cm.pretty_format(sort_by_count=True, show_percents=True))
print 'The accuracy is:',round(float(cm_sk[1][1]+cm_sk[0][0])/float(sum(cm_sk)),3)

     |      H        |
     |      i      L |
     |      g      o |
     |      h      w |
-----+---------------+
High | <54.1%> 14.6% |
 Low |  14.2% <17.2%>|
-----+---------------+
(row = reference; col = test)

The accuracy is: 0.712


#### Task E. Implement the PMI approach to sentiment analysis (in either Python or R), and run the classification model with the sentiment scores. How do your results compare with those in Task D?

In [37]:
import functools
from nltk.corpus import stopwords as stp
from nltk import pos_tag
import requests
from bs4 import BeautifulSoup

In [31]:
def features_extraction(text, rule_1, rule_2, no_match):
    
    wc = []
    append = wc.append
    text = text.lower()
    tokens = re.split("\W+", text)
    stopwords = stp.words("english")
    remove = tokens.remove
    while "" in tokens:
        remove("")
    for token in tokens:
        if token in stopwords:
            remove(token)
    token_pos = pos_tag(tokens)
    
    for i in xrange(len(token_pos)-1):
        if (token_pos[i][1], token_pos[i+1][1]) not in rule_1 and\
            (token_pos[i][1], token_pos[i+1][1]) not in rule_2:
                continue
        elif (token_pos[i][1], token_pos[i+1][1]) in rule_1:
            append((token_pos[i][0], token_pos[i+1][0]))
        elif (token_pos[i][1], token_pos[i+1][1]) in rule_2:
            try:
                if token_pos[i+2] not in no_match:
                    append((token_pos[i][0], token_pos[i+1][0]))
            except IndexError:
                append((token_pos[i][0], token_pos[i+1][0]))
        else:
            continue

    return wc

In [32]:
def Semantic_Orientation(phrases,positive="excellent", negative="poor", prior = 1,engine="google",distance=10,threshold=None):
  
    so_positive = 0.01
    so_negative = 0.01
    so_avg = 0
    get = requests.get

    engine.lower() == "google"
    url = "http://www.google.com/search?q=%s"
    id_pattern = "resultStats"
    
    for phrase in phrases:
        term = "%22{}+{}%22+AROUND({})+%22{}%22".format(phrase[0], distance, phrase[1], positive)
        page = get(url % term)
        soup = BeautifulSoup(page.text, "lxml")
        try:
            rtr_pos = int("".join(re.split("\D+",soup.find("div", id=id_pattern).get_text().encode("utf-８"))))
        except AttributeError:
            rtr_pos = 0
            
        # if rtr_pos == None:
        #     rtr_pos = 0
        
        term = "%22{}+{}%22+AROUND({})+%22{}%22".format(phrase[0], distance, phrase[1], negative)
        page = get(url % term)
        soup = BeautifulSoup(page.text, "lxml")
        try:
            rtr_neg = int("".join(re.split("\D+",soup.find("div", id=id_pattern).get_text().encode("utf-８"))))
        except AttributeError:
            rtr_neg = 0
            
        # if rtr_neg == None:
         #   rtr_neg =0
            
        so_positive += rtr_pos
        so_negative += rtr_neg
        
        so_avg += math.log(so_positive/so_negative/prior)
    
    #if len(phrases)==0:
     #   so_avg=0
    #else:
        so_avg = so_avg*1.0/len(phrases)
    
    if threshold == None:
        return so_avg
    else:
        return int(so_avg>threshold)

In [40]:
# Defining Patters for Feature Extraction
rule_1 = [("JJ", "NN"), ("JJ", "NNS"), ("RB", "VB"), ("RB", "VBD"), ("RB", "VBN"), ("RB", "VBG"),("RBR", "VB"), ("RBR", "VBD"), ("RBR", "VBN"), ("RBR", "VBG"),
          ("RBS", "VB"), ("RBS", "VBD"), ("RBS", "VBN"), ("RBS", "VBG")]
rule_2 = [("RB", "JJ"), ("RBR", "JJ"), ("RBS", "JJ"),("JJ", "JJ"),("NN", "JJ"), ("NNS", "JJ")]
no_match = ["NN", "NNS"]

# Extracting Features

features_extract = functools.partial(features_extraction, rule_1=rule_1, rule_2=rule_2, no_match=no_match)
features=yelp_subset.Review.map(features_extract)

In [41]:
# Calling the function that gives Semantic Orientation for a review.

Opinion_Orientation = functools.partial(Semantic_Orientation,positive="excellent", negative="poor",prior = 1,engine="google",distance=5,threshold=None)

# Mapping the function to our review list(sample :12)

prediction = features[:12].map(Opinion_Orientation)

ValueError: invalid literal for int() with base 10: ''

In [44]:
# Checking accuracy
y_pred = prediction.tolist()
y_true = y[:12].tolist() 


cm_sk = confusion_matrix(y_true, y_pred)
cm = nltk.ConfusionMatrix(y_true, y_pred)


print (cm.pretty_format(sort_by_count=True, show_percents=True))
if cm_sk[1][1] ==None:
    sum_sk[1][1]=0;
print 'The accuracy is:',round(float(cm_sk[1][1]+cm_sk[0][0])/float(len(y_pred)),3)

NameError: name 'prediction' is not defined

#### Task F. What are the top 5 “attributes” of a restaurant that are associated with (i) high and (ii) low ratings? That is, when people rate a restaurant high or low, are they more likely to mention service, ambiance, etc.? 

In [45]:
from sklearn.decomposition import TruncatedSVD
from sklearn.preprocessing import Normalizer
from sklearn import metrics
from sklearn.cluster import KMeans, MiniBatchKMeans

In [66]:
pos_reviews  = yelp.ix[yelp.rating == 'High']['Review']
vectorizer = CountVectorizer(min_df = 0, stop_words = 'english')
dtm = vectorizer.fit_transform(pos_reviews)

In [67]:
#Perform Latent Semantic Analysis on the matrix with 5 components
lsa = TruncatedSVD(5, algorithm = 'randomized')
dtm_lsa = lsa.fit_transform(dtm)
dtm_lsa = Normalizer(copy=False).fit_transform(dtm_lsa)
pos_lsa = pd.DataFrame(dtm_lsa, index = pos_reviews, columns = ["component_1","component_2","component_3","component_4","component_5" ])

##### The reviews are ranked based on the score in individual components, then topics are extracted from the top reviews.

component 1: people tend to talk about the service

In [68]:
pos_lsa.sort_values('component_1', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
absolutely delicious indian fare. can even get ...     0.998425    -0.019216   
Oh man  am I the only one who loves Macayo s?? ...     0.996507    -0.070234   
This little place does a great job pumping out ...     0.996463    -0.004482   
The Windsor drips with attitude and atmosphere....     0.996108    -0.006882   
Wow.  I am not a fan of casual bar type restaur...     0.995648     0.042703   
I adore El Bravo!I admit that I may be biased b...     0.995575     0.044998   
We showed up at about 4:55pm on a Friday  so th...     0.994701    -0.024211   
I have been to Chelsea s more than 10 times  an...     0.994700    -0.076628   
Okay  how can you not like Texaz Grill? Bad moo...     0.994438    -0.049122   
My husband and I decided to go to lunch on Mond...     0.994360     0.029902   
Supa Zupas! Ok enough of that. Went the other d...     0.993519    -0.058988   
I scored a date a few nights ago with a really ...     0.993464     0.005402   
China Village  I would like to award you 5 star...     0.992774     0.089596   
You know how some mornings you just wake up hun...     0.992682     0.018690   
I went to this little place for the first time ...     0.992429     0.017206   
The Good:  :: the sushi chef (Yasu). he knows h...     0.992401     0.075061   
I was fooled.  The unassuming strip mall exteri...     0.992001     0.056580   
The following is written by me  one of Richards...     0.991944    -0.030398   
Z Tejas is quickly becoming one of our top  go ...     0.991237    -0.023083   
If you re in D concourse in PHX  and you re hun...     0.990882     0.085703   

                                                    component_3  component_4  \
Review                                                                         
absolutely delicious indian fare. can even get ...    -0.038637    -0.029912   
Oh man  am I the only one who loves Macayo s?? ...    -0.024420     0.027169   
This little place does a great job pumping out ...    -0.006548     0.075198   
The Windsor drips with attitude and atmosphere....    -0.049519    -0.029546   
Wow.  I am not a fan of casual bar type restaur...     0.048930    -0.059183   
I adore El Bravo!I admit that I may be biased b...     0.048481    -0.050675   
We showed up at about 4:55pm on a Friday  so th...     0.005280    -0.070130   
I have been to Chelsea s more than 10 times  an...    -0.051661     0.004346   
Okay  how can you not like Texaz Grill? Bad moo...     0.062551     0.057316   
My husband and I decided to go to lunch on Mond...     0.035297     0.095324   
Supa Zupas! Ok enough of that. Went the other d...     0.005114    -0.083657   
I scored a date a few nights ago with a really ...    -0.016628    -0.078429   
China Village  I would like to award you 5 star...    -0.067840     0.010644   
You know how some mornings you just wake up hun...    -0.076492    -0.073543   
I went to this little place for the first time ...    -0.046674    -0.066672   
The Good:  :: the sushi chef (Yasu). he knows h...    -0.089486     0.031525   
I was fooled.  The unassuming strip mall exteri...     0.002253     0.002792   
The following is written by me  one of Richards...    -0.029451    -0.118825   
Z Tejas is quickly becoming one of our top  go ...     0.083778     0.013301   
If you re in D concourse in PHX  and you re hun...     0.089870     0.052251   

                                                    component_5  
Review                                                           
absolutely delicious indian fare. can even get ...    -0.019772  
Oh man  am I the only one who loves Macayo s?? ...     0.026563  
This little place does a great job pumping out ...    -0.036665  
The Windsor drips with attitude and atmosphere....    -0.066302  
Wow.  I am not a fan of casual bar type restaur...    -0.031050  
I adore El Bravo!I admit that I may be biased b...    -0.043445  
We

component 2: people like to talk about the food quality

In [69]:
pos_lsa.sort_values('component_2', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
Zupa di pesch was phenom!                              0.406261     0.708828   
YUM  if you like a pizza with no sauce  get the...     0.634075     0.700003   
Hang on...gotta wipe the cheese off my chin.  O...     0.609345     0.684888   
Butterfish and chocolate souffle are pure heaven!      0.459061     0.682092   
what is there to be said that hasn t already be...     0.668425     0.674634   
I ve been here dozens and DOZENS of times over ...     0.693850     0.673816   
Skip the full-sized burger and order sliders; m...     0.667710     0.672227   
Wow..got a $5.00 for just for checking in..            0.682862     0.663885   
Pizza is spicy and sweet.  Lovely                      0.621319     0.654324   
YUM!Maybe it s just because we had extra sauce ...     0.696479     0.635730   
No better slice of cheese pizza in town. But on...     0.700461     0.634204   
Have you tried their  banana coconut cream pie?...     0.650417     0.632831   
I don t know if it s because I was starving  bu...     0.704960     0.624031   
The bean and cheese burritos are made of crack....     0.585445     0.622179   
Yay  they re back!! Now to protect my kingdom :)       0.637681     0.621159   
I was fixated on the Rocket  Burger  of Rocket ...     0.684435     0.615635   
Butter Cake                                            0.418244     0.614411   
Two words- Smokey Forest.                              0.673933     0.613202   
Cheddar burger and fire dogs - awesome.  Burger...     0.696917     0.613173   
Upgrading to 3.75JT s menu seems to expand ever...     0.734178     0.611477   

                                                    component_3  component_4  \
Review                                                                         
Zupa di pesch was phenom!                             -0.460554     0.094817   
YUM  if you like a pizza with no sauce  get the...    -0.225270     0.084178   
Hang on...gotta wipe the cheese off my chin.  O...    -0.236577     0.038742   
Butterfish and chocolate souffle are pure heaven!     -0.285400    -0.064395   
what is there to be said that hasn t already be...    -0.250695     0.065184   
I ve been here dozens and DOZENS of times over ...    -0.243373    -0.012165   
Skip the full-sized burger and order sliders; m...    -0.287045    -0.092147   
Wow..got a $5.00 for just for checking in..           -0.280125     0.038975   
Pizza is spicy and sweet.  Lovely                     -0.274444    -0.287963   
YUM!Maybe it s just because we had extra sauce ...    -0.310020     0.010144   
No better slice of cheese pizza in town. But on...    -0.293174    -0.124323   
Have you tried their  banana coconut cream pie?...    -0.288464    -0.036889   
I don t know if it s because I was starving  bu...    -0.164781     0.094585   
The bean and cheese burritos are made of crack....    -0.190884     0.242067   
Yay  they re back!! Now to protect my kingdom :)      -0.368653    -0.235144   
I was fixated on the Rocket  Burger  of Rocket ...    -0.295135    -0.219999   
Butter Cake                                           -0.270687    -0.022469   
Two words- Smokey Forest.                             -0.215600     0.329094   
Cheddar burger and fire dogs - awesome.  Burger...    -0.322742    -0.166114   
Upgrading to 3.75JT s menu seems to expand ever...    -0.227404     0.145166   

                                                    component_5  
Review                                                           
Zupa di pesch was phenom!                             -0.333787  
YUM  if you like a pizza with no sauce  get the...     0.223858  
Hang on...gotta wipe the cheese off my chin.  O...    -0.319620  
Butterfish and chocolate souffle are pure heaven!     -0.488277  
what is there to be said that hasn t already be...     0.176014  
I ve been here dozens and DOZENS of times over ...    -0.071864  
Sk

component 3: people give have a good overall impression

In [71]:
pos_lsa.sort_values('component_3', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
So good                                                0.337852     0.112690   
Very good                                              0.337852     0.112690   
ridiculously good ...                                  0.340530     0.113996   
Good and unique.                                       0.351827     0.111250   
Good tofu!                                             0.349333     0.117313   
Good gyro meat. Tzatziki is good.                      0.360123     0.127482   
Never been to the sky harbor location . The one...     0.396344     0.105429   
I had Filet Mignon as usual and it was not as g...     0.360775     0.131982   
I recommend the Pollo Siciliano and the Pollo b...     0.385702     0.109531   
Very good cornbread and the street tacos were a...     0.392198     0.112550   
Outstanding and fantastalistic comes to mind. R...     0.400467     0.131914   
messy to eat----but ohhhh so good                      0.401894     0.128408   
Always good times at Mi patio..Cheap margaritas...     0.384907     0.105853   
Good and I WILL go to this diner again. It gets...     0.397669     0.118313   
Fish tacos are good.                                   0.401180     0.123043   
This places ooozes good taste                          0.412075     0.136016   
#3 was very good  all freshly cut meat.                0.390682     0.162587   
Ridiculously tasty. Not good for you.                  0.388928     0.146739   
Durant s reminds me of good lovin . Not your ev...     0.399181     0.136198   
If it sounds good  get it. My friends and I hav...     0.394098     0.118272   

                                                    component_3  component_4  \
Review                                                                         
So good                                                0.894676    -0.228117   
Very good                                              0.894676    -0.228117   
ridiculously good ...                                  0.894337    -0.225026   
Good and unique.                                       0.889490    -0.227309   
Good tofu!                                             0.889022    -0.224999   
Good gyro meat. Tzatziki is good.                      0.886213    -0.221036   
Never been to the sky harbor location . The one...     0.880571    -0.193839   
I had Filet Mignon as usual and it was not as g...     0.880121    -0.232126   
I recommend the Pollo Siciliano and the Pollo b...     0.879443    -0.219770   
Very good cornbread and the street tacos were a...     0.878549    -0.197531   
Outstanding and fantastalistic comes to mind. R...     0.875422    -0.197918   
messy to eat----but ohhhh so good                      0.875393    -0.211943   
Always good times at Mi patio..Cheap margaritas...     0.875348    -0.229276   
Good and I WILL go to this diner again. It gets...     0.872954    -0.214095   
Fish tacos are good.                                   0.871682    -0.195936   
This places ooozes good taste                          0.870628    -0.208801   
#3 was very good  all freshly cut meat.                0.869602    -0.208630   
Ridiculously tasty. Not good for you.                  0.868983    -0.207921   
Durant s reminds me of good lovin . Not your ev...     0.868982    -0.215552   
If it sounds good  get it. My friends and I hav...     0.868567    -0.234738   

                                                    component_5  
Review                                                           
So good                                               -0.143785  
Very good                                             -0.143785  
ridiculously good ...                                 -0.143417  
Good and unique.                                      -0.144844  
Good tofu!                                            -0.152380  
Good gyro meat. Tzatziki is good.                     -0.140814  
Ne

component 4: a lot of mexican food

In [72]:
pos_lsa.sort_values('component_4', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
Chiptole.... enough saidgreat mexican for being...     0.344643    -0.364541   
This is how Mexican food should be served:It sh...     0.477516    -0.389773   
TOR.TI.LLAS......Delicioso :)!! Andale mi amgos...     0.547817    -0.281677   
spicy new mexican! yum!                                0.598091    -0.167737   
The fluffiest rice I have ever eaten!                  0.617523     0.176749   
Hushpuppies and gumbo!                                 0.468650    -0.041797   
One of our favorite Mexican food restaurants in...     0.574452    -0.323384   
I Heart Los!!!This is my absolute favorite Mexi...     0.646511    -0.021358   
Ben B. and I visited Al-Hana after a tiresome I...     0.568055    -0.353513   
The only spicy Mexican food I ve had in the val...     0.457270    -0.496075   
I like fast Mexican food and out of all the cor...     0.590524    -0.275299   
Had the buffet.  Loved it!  Food was hot but no...     0.588686    -0.272364   
BEST Mexican food restaurant this side of the M...     0.471192    -0.506662   
amazing food   the best ever . I have tried shi...     0.554113    -0.415473   
From LA.  Fast food Mexican is no better than t...     0.404177    -0.549282   
Yes this deserves a five star review. Everytime...     0.579309    -0.371061   
Muy mas caliente  but my favorite restaurant in...     0.638858    -0.232577   
Dive Restaurant with Incredible Authentic Mexic...     0.476804    -0.508286   
Reading some of these reviews do not shock me.....     0.678387    -0.136023   
Chicken chimayo yum yum!!Huge portions.... Sooo...     0.495784    -0.477830   

                                                    component_3  component_4  \
Review                                                                         
Chiptole.... enough saidgreat mexican for being...     0.139623     0.802529   
This is how Mexican food should be served:It sh...     0.091255     0.778591   
TOR.TI.LLAS......Delicioso :)!! Andale mi amgos...     0.038531     0.773673   
spicy new mexican! yum!                               -0.058211     0.768364   
The fluffiest rice I have ever eaten!                 -0.009683     0.765245   
Hushpuppies and gumbo!                                 0.150965     0.755563   
One of our favorite Mexican food restaurants in...     0.009443     0.749321   
I Heart Los!!!This is my absolute favorite Mexi...    -0.139500     0.747998   
Ben B. and I visited Al-Hana after a tiresome I...     0.041101     0.737383   
The only spicy Mexican food I ve had in the val...     0.094114     0.729860   
I like fast Mexican food and out of all the cor...     0.014713     0.723798   
Had the buffet.  Loved it!  Food was hot but no...     0.050618     0.723685   
BEST Mexican food restaurant this side of the M...     0.044473     0.720592   
amazing food   the best ever . I have tried shi...     0.043182     0.716170   
From LA.  Fast food Mexican is no better than t...     0.147374     0.715482   
Yes this deserves a five star review. Everytime...     0.049115     0.714361   
Muy mas caliente  but my favorite restaurant in...    -0.064739     0.711941   
Dive Restaurant with Incredible Authentic Mexic...     0.062864     0.711849   
Reading some of these reviews do not shock me.....    -0.071627     0.710138   
Chicken chimayo yum yum!!Huge portions.... Sooo...     0.096238     0.708456   

                                                    component_5  
Review                                                           
Chiptole.... enough saidgreat mexican for being...     0.291177  
This is how Mexican food should be served:It sh...     0.074328  
TOR.TI.LLAS......Delicioso :)!! Andale mi amgos...     0.143181  
spicy new mexican! yum!                                0.142760  
The fluffiest rice I have ever eaten!                 -0.041608  
Hushpuppies and gumbo!                                -0.430063  
On

component 5: The overall place, such as atmosphere

In [73]:
pos_lsa.sort_values('component_5', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
Everyone I have told about this place is like  ...     0.468238    -0.049067   
I would like to know what happen to Coffee Dans...     0.530397     0.042457   
May look like a dump I a bad neighborhood  but ...     0.569175    -0.103344   
The absolute best!!  I love this place and have...     0.560613     0.043147   
Oye  que tal  mis amigos? This is the premier p...     0.538186    -0.085427   
Bomba! Wow.. I m here eating while I draft this...     0.431159    -0.227928   
Sometimes you have to explore and try those hol...     0.577348    -0.144812   
are you joking? Where else can you get the best...     0.554736    -0.097604   
Trust the positive reviews...   This place is l...     0.515281    -0.161290   
I m amazed that this place is in Phoenix. It ca...     0.586001     0.004115   
Because sometimes you just want to flip the bir...     0.523442    -0.118302   
So nothing is perfect  but for a guy who loves ...     0.491596    -0.109144   
Order in  order for take-out and pick up at the...     0.528689    -0.170202   
This place is better than mostGyros are the bes...     0.359867    -0.309922   
i have been coming to this location over 15 yea...     0.573019    -0.049910   
This place is never disappointing!                     0.310742    -0.337141   
Make sure you pull into this place caddy corner...     0.468993    -0.327943   
If i had to pick one place to go for my birthda...     0.351228    -0.310222   
It was silly to only give this legendary place ...     0.343705    -0.329501   
Bitchin place to refuel after shreddin Desert C...     0.371924    -0.334356   

                                                    component_3  component_4  \
Review                                                                         
Everyone I have told about this place is like  ...    -0.119314    -0.221381   
I would like to know what happen to Coffee Dans...    -0.159036    -0.204756   
May look like a dump I a bad neighborhood  but ...    -0.122710    -0.080772   
The absolute best!!  I love this place and have...    -0.191521    -0.073261   
Oye  que tal  mis amigos? This is the premier p...    -0.146785    -0.207487   
Bomba! Wow.. I m here eating while I draft this...    -0.109053    -0.359530   
Sometimes you have to explore and try those hol...    -0.143110    -0.174413   
are you joking? Where else can you get the best...    -0.165878    -0.250172   
Trust the positive reviews...   This place is l...    -0.150112    -0.306597   
I m amazed that this place is in Phoenix. It ca...    -0.202664    -0.157330   
Because sometimes you just want to flip the bir...    -0.178397    -0.307076   
So nothing is perfect  but for a guy who loves ...    -0.149495    -0.373495   
Order in  order for take-out and pick up at the...    -0.180819    -0.275154   
This place is better than mostGyros are the bes...    -0.067524    -0.433476   
i have been coming to this location over 15 yea...    -0.195394    -0.221678   
This place is never disappointing!                    -0.057529    -0.454576   
Make sure you pull into this place caddy corner...    -0.111911    -0.283531   
If i had to pick one place to go for my birthda...    -0.084966    -0.441258   
It was silly to only give this legendary place ...    -0.059464    -0.438618   
Bitchin place to refuel after shreddin Desert C...    -0.048345    -0.413749   

                                                    component_5  
Review                                                           
Everyone I have told about this place is like  ...     0.845636  
I would like to know what happen to Coffee Dans...     0.806014  
May look like a dump I a bad neighborhood  but ...     0.802358  
The absolute best!!  I love this place and have...     0.801127  
Oye  que tal  mis amigos? This is the premier p...     0.799038  
Bomba! Wow.. I m here eating while I draft this...     0.788033  
So

##### perform the same analysis on negative reviews

In [74]:
neg_reviews  = yelp.ix[yelp.rating == 'Low']['Review']
neg_dtm = vectorizer.fit_transform(neg_reviews)
neg_dtm_lsa = lsa.fit_transform(neg_dtm)
neg_dtm_lsa = Normalizer(copy=False).fit_transform(neg_dtm_lsa)
neg_lsa = pd.DataFrame(neg_dtm_lsa, index = neg_reviews, columns = ["component_1","component_2","component_3","component_4","component_5" ])

component 1: taste

In [75]:
neg_lsa.sort_values('component_1', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
I went to the Phoenix Hard Rock Cafe for lunch ...     0.997757     0.035643   
Decent sushi  I prefer to sit up at the bar cou...     0.997702    -0.009090   
Beats Burger King any day. I should ve manned u...     0.997115    -0.060117   
Bocce Ball s  that was a some spicy Tortillion!...     0.996934    -0.031109   
Unfortunately  I was slightly let down by the y...     0.995879    -0.041258   
Eh... pasta is way to salty.  Just just cause M...     0.995781    -0.042049   
The food was pretty tasty  but it s definitely ...     0.993996     0.033765   
How many Uzbeki Jewish restaurants have you bee...     0.993440     0.071694   
I really wanted to love this place. I had initi...     0.993432    -0.064643   
Cranky bar staff and antiquated menu make this ...     0.993411    -0.079220   
So.... this place is ridiculously overpriced. I...     0.993162     0.096782   
It s dark in there!Honestly  I walked in last S...     0.992061    -0.073791   
With the exception of very friendly employees  ...     0.991389     0.086864   
To start  I ll let the facts speak for themselv...     0.991372     0.037346   
I figured this place must have been good  since...     0.990274    -0.130761   
Got a half brisket sandwich and salad at lunch ...     0.990044     0.025978   
Meh. Service was not very quick but it was frie...     0.989639    -0.029050   
Meh.  Just meh.  OK - maybe Dim Sum is somethin...     0.989625    -0.036450   
I was visiting a friend who just moved here  ev...     0.989451    -0.001732   
AJ s has become one of my Sunday lunch favorite...     0.989298     0.085366   

                                                    component_3  component_4  \
Review                                                                         
I went to the Phoenix Hard Rock Cafe for lunch ...     0.013533     0.032991   
Decent sushi  I prefer to sit up at the bar cou...    -0.046845    -0.030324   
Beats Burger King any day. I should ve manned u...     0.043692    -0.015413   
Bocce Ball s  that was a some spicy Tortillion!...    -0.068913     0.019784   
Unfortunately  I was slightly let down by the y...    -0.045082    -0.023923   
Eh... pasta is way to salty.  Just just cause M...    -0.070677     0.027326   
The food was pretty tasty  but it s definitely ...    -0.001269    -0.102057   
How many Uzbeki Jewish restaurants have you bee...     0.071505    -0.032664   
I really wanted to love this place. I had initi...     0.057194    -0.037208   
Cranky bar staff and antiquated menu make this ...    -0.006458     0.048527   
So.... this place is ridiculously overpriced. I...    -0.013059    -0.041401   
It s dark in there!Honestly  I walked in last S...     0.080869    -0.041774   
With the exception of very friendly employees  ...     0.048471    -0.083752   
To start  I ll let the facts speak for themselv...    -0.016884    -0.017246   
I figured this place must have been good  since...    -0.025644     0.016537   
Got a half brisket sandwich and salad at lunch ...     0.123258     0.047559   
Meh. Service was not very quick but it was frie...     0.109322     0.060295   
Meh.  Just meh.  OK - maybe Dim Sum is somethin...    -0.063741     0.116076   
I was visiting a friend who just moved here  ev...    -0.072540    -0.030813   
AJ s has become one of my Sunday lunch favorite...     0.087121     0.020508   

                                                    component_5  
Review                                                           
I went to the Phoenix Hard Rock Cafe for lunch ...     0.044041  
Decent sushi  I prefer to sit up at the bar cou...    -0.037325  
Beats Burger King any day. I should ve manned u...     0.000909  
Bocce Ball s  that was a some spicy Tortillion!...     0.003909  
Unfortunately  I was slightly let down by the y...    -0.062592  
Eh... pasta is way to salty.  Just just cause M...    -0.030164  
Th

component 2: bad food quality

In [76]:
neg_lsa.sort_values('component_2', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
Americanized Japanese food.                            0.382128     0.878643   
Grotesque food  vulgar staff  blech.                   0.406268     0.877840   
Over priced and mediocre food.                         0.393311     0.875812   
Excellent food as always                               0.389897     0.875287   
Wish the food was hotter.                              0.394472     0.875021   
Food is average. The hours are ridiculous! Its ...     0.409232     0.873912   
It s Pei Wei. The service and food are always s...     0.436014     0.872509   
Decent food.                                           0.404465     0.870786   
Incompetent staff. Medicore food. Not impressed...     0.427382     0.870447   
Cheap as can be. The coffee is terrible  the fo...     0.419494     0.867567   
average food and service in my opinion.  I eat ...     0.441928     0.867118   
This is not an Indian food place. This is a fau...     0.450185     0.864652   
The staff was rude and unfriendly.  The music w...     0.462231     0.864203   
Snooty and expensive. Tiny portions of  preciou...     0.418462     0.862740   
Horrible disgusting food will never eat here ag...     0.426188     0.861895   
Nasty. Terrible service  hotel food. Save your ...     0.470297     0.859052   
Is this mexican food?                                  0.370855     0.856945   
Service is fine. quality of food has fallen sig...     0.469788     0.855893   
Some over priced Mexican food in the airport.          0.379273     0.853670   
Terrible service  inapproite behavior by multip...     0.497556     0.850778   

                                                    component_3  component_4  \
Review                                                                         
Americanized Japanese food.                           -0.096958     0.030165   
Grotesque food  vulgar staff  blech.                  -0.086511     0.035543   
Over priced and mediocre food.                        -0.096532     0.028194   
Excellent food as always                              -0.088606     0.050488   
Wish the food was hotter.                             -0.087446     0.033354   
Food is average. The hours are ridiculous! Its ...    -0.076899     0.026667   
It s Pei Wei. The service and food are always s...     0.012832     0.182656   
Decent food.                                          -0.108123     0.033767   
Incompetent staff. Medicore food. Not impressed...    -0.092096     0.038735   
Cheap as can be. The coffee is terrible  the fo...    -0.093684     0.012130   
average food and service in my opinion.  I eat ...    -0.079490     0.066288   
This is not an Indian food place. This is a fau...    -0.160578    -0.124063   
The staff was rude and unfriendly.  The music w...    -0.060979     0.044433   
Snooty and expensive. Tiny portions of  preciou...    -0.114338     0.020786   
Horrible disgusting food will never eat here ag...    -0.064635     0.010824   
Nasty. Terrible service  hotel food. Save your ...     0.029417     0.182858   
Is this mexican food?                                 -0.156137    -0.033137   
Service is fine. quality of food has fallen sig...    -0.001354     0.198652   
Some over priced Mexican food in the airport.         -0.160349    -0.032559   
Terrible service  inapproite behavior by multip...     0.010398     0.154084   

                                                    component_5  
Review                                                           
Americanized Japanese food.                            0.267683  
Grotesque food  vulgar staff  blech.                   0.235790  
Over priced and mediocre food.                         0.261049  
Excellent food as always                               0.267308  
Wish the food was hotter.                              0.264521  
Food is average. The hours are ridiculous! Its ...     0.249365  
It

component 3: bad service

In [77]:
neg_lsa.sort_values('component_3', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
It s too confusing to order here.Makes me get a...     0.482915    -0.048767   
Employees gathered in the back chatting so empt...     0.565244    -0.120123   
While I was eating  a coach roach ran across my...     0.546275    -0.050501   
No!I said - NO!!!                                      0.442284    -0.299590   
Waitress took a long time to seat us. Then didn...     0.540576    -0.141385   
Wildly uneven depending on who s in the kitchen...     0.642815     0.010003   
disappointing carnitas burrito. no kick in the ...     0.568316    -0.129510   
They have changed their recipes and several ite...     0.598604    -0.096650   
Walked in first before two other couples came i...     0.615604    -0.225340   
They treat their coustomers so bad that once I ...     0.628196    -0.029710   
OUT OF BUSINESS WHEN I STOPPED BY ON 12/15/10          0.530311     0.028708   
Went there for breakfast. Only two people  I th...     0.548132    -0.000810   
2 stars for free pizza. -3 stars for comedicall...     0.541281    -0.166288   
We got there and told them we had a party of 8 ...     0.634770    -0.157494   
I used to be a customer....even tho the fries a...     0.636539    -0.275187   
I made a reservation for Donovan s prix fixe me...     0.641821    -0.188579   
The first time I went there and sat at the coun...     0.611627    -0.139295   
One thing that I love about OSF is that they of...     0.631073    -0.132205   
Mother s day wrecked.  Thanks Rico s.  I called...     0.614871    -0.046890   
Western Bacon Cheeseburger. Enough said.               0.511403    -0.371493   

                                                    component_3  component_4  \
Review                                                                         
It s too confusing to order here.Makes me get a...     0.818747     0.300829   
Employees gathered in the back chatting so empt...     0.784008     0.226621   
While I was eating  a coach roach ran across my...     0.779597     0.289890   
No!I said - NO!!!                                      0.779129     0.319658   
Waitress took a long time to seat us. Then didn...     0.765010     0.303465   
Wildly uneven depending on who s in the kitchen...     0.761896     0.064751   
disappointing carnitas burrito. no kick in the ...     0.742571     0.329820   
They have changed their recipes and several ite...     0.742082     0.285638   
Walked in first before two other couples came i...     0.737184     0.151225   
They treat their coustomers so bad that once I ...     0.732816     0.224926   
OUT OF BUSINESS WHEN I STOPPED BY ON 12/15/10          0.729694     0.401056   
Went there for breakfast. Only two people  I th...     0.727962     0.326440   
2 stars for free pizza. -3 stars for comedicall...     0.726132     0.309984   
We got there and told them we had a party of 8 ...     0.724945     0.215921   
I used to be a customer....even tho the fries a...     0.719001     0.036166   
I made a reservation for Donovan s prix fixe me...     0.714849     0.194602   
The first time I went there and sat at the coun...     0.710679     0.253709   
One thing that I love about OSF is that they of...     0.705412     0.255943   
Mother s day wrecked.  Thanks Rico s.  I called...     0.704322     0.311238   
Western Bacon Cheeseburger. Enough said.               0.702276     0.325061   

                                                    component_5  
Review                                                           
It s too confusing to order here.Makes me get a...     0.059757  
Employees gathered in the back chatting so empt...     0.006661  
While I was eating  a coach roach ran across my...    -0.085001  
No!I said - NO!!!                                      0.073534  
Waitress took a long time to seat us. Then didn...    -0.102254  
Wildly uneven depending on who s in the kitchen...    -0.044848  
di

component 4: many think the food is pricy

In [78]:
neg_lsa.sort_values('component_4', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
Solid.  In and out in 45 minutes with a full st...     0.539128    -0.030728   
Not very good service. Orders came out wrong. W...     0.611523    -0.036162   
Good service bbq was tastee. Overall happy!            0.467456    -0.066058   
Good sandwiches and quick service.                     0.447463    -0.038852   
Not bad.  Good wait staff.  Quick service.  She...     0.542445     0.015040   
Appetizer platter was large and tasty.The galak...     0.613660    -0.130006   
So the burgers are good  as most people on Yelp...     0.710411    -0.096757   
Here for conference.  Convenient location  attr...     0.551209    -0.147289   
Service was good even though they messed up my ...     0.579242    -0.225967   
Went for my Bday dinner. Nachos were good  marg...     0.495660    -0.276524   
Went for dinner with a friend.  Vegetable Delux...     0.613749    -0.158321   
We had a pleasant time for Aloha here on Monday...     0.609834    -0.164723   
As loyal Richardsons fans  we went to the Roker...     0.757241     0.120474   
Love other locations  this one-NOT SO MUCH.Took...     0.598873    -0.037941   
I ordered a salmon taco which was delicious but...     0.547710    -0.268745   
Food was good  service was lousy. Started out w...     0.650386     0.325692   
Went with bf on a week night for a nice steak d...     0.698656    -0.284500   
I came here on Cinco de Mayo and it was oddly a...     0.693783    -0.148960   
Inexpensive and good service.  Will go back and...     0.587453    -0.161562   
First off  decent service if you need to grab s...     0.701987     0.072555   

                                                    component_3  component_4  \
Review                                                                         
Solid.  In and out in 45 minutes with a full st...    -0.290244     0.752989   
Not very good service. Orders came out wrong. W...     0.064605     0.738841   
Good service bbq was tastee. Overall happy!           -0.494718     0.692860   
Good sandwiches and quick service.                    -0.523953     0.680662   
Not bad.  Good wait staff.  Quick service.  She...    -0.403809     0.667522   
Appetizer platter was large and tasty.The galak...    -0.381581     0.661071   
So the burgers are good  as most people on Yelp...    -0.113392     0.659494   
Here for conference.  Convenient location  attr...    -0.459353     0.654763   
Service was good even though they messed up my ...    -0.417885     0.653313   
Went for my Bday dinner. Nachos were good  marg...    -0.500851     0.646234   
Went for dinner with a friend.  Vegetable Delux...    -0.372363     0.637883   
We had a pleasant time for Aloha here on Monday...     0.399607     0.634078   
As loyal Richardsons fans  we went to the Roker...     0.072227     0.633574   
Love other locations  this one-NOT SO MUCH.Took...     0.407144     0.631977   
I ordered a salmon taco which was delicious but...    -0.481040     0.629389   
Food was good  service was lousy. Started out w...    -0.276386     0.627136   
Went with bf on a week night for a nice steak d...     0.194879     0.626704   
I came here on Cinco de Mayo and it was oddly a...    -0.269590     0.626121   
Inexpensive and good service.  Will go back and...    -0.459670     0.624489   
First off  decent service if you need to grab s...    -0.302292     0.623982   

                                                    component_5  
Review                                                           
Solid.  In and out in 45 minutes with a full st...    -0.239088  
Not very good service. Orders came out wrong. W...    -0.273261  
Good service bbq was tastee. Overall happy!           -0.228735  
Good sandwiches and quick service.                    -0.245846  
Not bad.  Good wait staff.  Quick service.  She...    -0.311256  
Appetizer platter was large and tasty.The galak...    -0.154600  
So

In [79]:
neg_lsa.sort_values('component_5', ascending=False).ix[:20]

component_1  component_2  \
Review                                                                         
Chicken fried rice Yum. Friendly employees. Cou...     0.314272    -0.210849   
The yougurt dipping sauce is delish!                   0.307492    -0.334664   
Sigh... the search for anything remotely asian ...     0.521498    -0.069653   
Took a feel bites of my Teriyaki Chicken and ha...     0.490743    -0.307515   
HORRIBLE!!!!    I ordered Chicken Fried Rice  G...     0.459390    -0.305909   
CervicheFish tacosCombo - taco                         0.409464     0.174739   
Avoid the lunch buffet.  Excellent Chicken Tikk...     0.428579    -0.215102   
I have tried the teriyaki chicken bowl and the ...     0.415324    -0.253351   
Ahi salad was poor....meat and dressing had no ...     0.453774    -0.398010   
Love the chicken broccoli w cheese yumm                0.405274    -0.400661   
Bad panda! Bad!Hot & Sour soup was actually so ...     0.474145    -0.307700   
Stopped in for lunch  and it was definitely a s...     0.470310    -0.256433   
Happy hour Margarita was better than average an...     0.576301    -0.281773   
Not a bad option if you are starving at the PHX...     0.621868    -0.222903   
carolinas  your enchiladas ruined it for me.i c...     0.478256    -0.234987   
I went with a party of four  and was the only o...     0.515747    -0.357432   
Glorified Pei Wei                                      0.205618     0.519488   
The Brown butter sauce and pesto ranch are both...     0.458048    -0.348308   
Disappointed  had the shrimp burrito  lots of r...     0.528295    -0.306432   
Fun restaurant for a night out with friends.  T...     0.557526    -0.299586   

                                                    component_3  component_4  \
Review                                                                         
Chicken fried rice Yum. Friendly employees. Cou...    -0.238715    -0.334074   
The yougurt dipping sauce is delish!                  -0.237798    -0.286713   
Sigh... the search for anything remotely asian ...    -0.190221    -0.225993   
Took a feel bites of my Teriyaki Chicken and ha...    -0.041139    -0.190566   
HORRIBLE!!!!    I ordered Chicken Fried Rice  G...    -0.085792    -0.266550   
CervicheFish tacosCombo - taco                        -0.299977    -0.319582   
Avoid the lunch buffet.  Excellent Chicken Tikk...    -0.266238    -0.310855   
I have tried the teriyaki chicken bowl and the ...    -0.209025    -0.353854   
Ahi salad was poor....meat and dressing had no ...    -0.181454    -0.130077   
Love the chicken broccoli w cheese yumm               -0.139680    -0.286843   
Bad panda! Bad!Hot & Sour soup was actually so ...    -0.177013    -0.282232   
Stopped in for lunch  and it was definitely a s...    -0.250979    -0.287220   
Happy hour Margarita was better than average an...    -0.061615    -0.162559   
Not a bad option if you are starving at the PHX...    -0.014604    -0.072658   
carolinas  your enchiladas ruined it for me.i c...    -0.218711    -0.336013   
I went with a party of four  and was the only o...    -0.221935    -0.043907   
Glorified Pei Wei                                      0.083170    -0.357323   
The Brown butter sauce and pesto ranch are both...    -0.161800    -0.313277   
Disappointed  had the shrimp burrito  lots of r...    -0.199059    -0.211858   
Fun restaurant for a night out with friends.  T...    -0.115359    -0.212553   

                                                    component_5  
Review                                                           
Chicken fried rice Yum. Friendly employees. Cou...     0.829570  
The yougurt dipping sauce is delish!                   0.809133  
Sigh... the search for anything remotely asian ...     0.797453  
Took a feel bites of my Teriyaki Chicken and ha...     0.791579  
HORRIBLE!!!!    I ordered Chicken Fried Rice  G...     0.785475  
CervicheFish tacosCombo - taco                         0.780824  
Av